h11/_headers.py

```diff
@@ -62,6 +68,28 @@
 _field_value_re = re.compile(field_value.encode("ascii"))
 
 
+class Headers(Sequence):
+    def __init__(self, items):
+        self._items = items
+
+    def __getitem__(self, item):
+        _, _, value = self._data[item]
+        return value
+
+    def __len__(self):
+        return len(self._items)
+
+    def __iter__(self):
+        for name, _, value in self._items:
+            yield name, value
+
+    def __eq__(self, other):
+        return list(self) == other
+
+    def raw(self):
```

Comment:
I'm curious about this. It seems it would return a list of the triplet where elsewhere `raw` implies the name that came over the wire. This should be clearer about what it is returning

Reply:
I also wonder about the wisdom of having a headers sequence without structured single-header data

Reply:
> I also wonder about the wisdom of having a headers sequence without structured single-header data

I'm not quite sure what you mean here. Are you saying "If you're returning a three-tuple from this interface then let's have it use a named-tuple" or something else?

Perhaps a marginally different interface for us to expose here would not be `.raw() -> (<lowercase name>, <raw name>, <value>)`, but instead expose just `.raw_items() -> (<raw name>, <value>)`

Perhaps that addressed the naming/intent slightly better?

Reply:
I'm thinking of having a class that encapsulates a `Header` and then having the collection of `Headers` use that. A namedtuple could work fine as well. All that said, I get that tuples may be a smidge faster and that these are internal implementation details. Speaking from having worked on header collection objects in urllib3 in the past, these tuples can drive maintainers to pull out their hair (as well as future folks trying to update/extend the behaviour).

Reply:
Right, for something that remains compatible with the existing API I think the options here are...

* A custom `Headers` sequence, that exposes the extra information in a `.raw_items()` interface or similar, that returns a two-tuple of `(case-sensitive-name, value)` for usages that require the raw casing info.
* A custom `Headers` sequence, the returns `Header` instances, that can iterate as two-tuples, but also expose `.name`, `.case_sensitive_name` and `.value` attributes, which are available for usages that require the raw casing info.

Or some variation on those. (Eg. this PR which currently has `.raw()` returning the three-tuple of info.)
Personally I'm fairly agnostic, as both the above options seem reasonable enough. The `Header` case has the most extra overhead, since it creates and accesses a per-header instance rather than the plain tuple, while I wouldn't expect the `.raw_items()` approach to introduce anything really noticeable, but I could run through some timings on each of the options to help better inform our options.